barretenberg/cpp/src/barretenberg/vm2/tracegen/bytecode_trace.cpp

```diff
@@ -246,95 +246,176 @@ void BytecodeTraceBuilder::process_instruction_fetching(
-                      { C::instr_fetching_bd36, bytecode_at(event.pc + 36) },
-
-                      // From instruction table.
-                      { C::instr_fetching_exec_opcode,
-                        static_cast<uint32_t>(WIRE_INSTRUCTION_SPEC.at(w_opcode).exec_opcode) },
-                      { C::instr_fetching_instr_size_in_bytes, WIRE_INSTRUCTION_SPEC.at(w_opcode).size_in_bytes },
-                      // Fill operand decomposition selectors
-                      { C::instr_fetching_sel_op_dc_0, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(0) },
-                      { C::instr_fetching_sel_op_dc_1, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(1) },
-                      { C::instr_fetching_sel_op_dc_2, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(2) },
-                      { C::instr_fetching_sel_op_dc_3, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(3) },
-                      { C::instr_fetching_sel_op_dc_4, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(4) },
-                      { C::instr_fetching_sel_op_dc_5, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(5) },
-                      { C::instr_fetching_sel_op_dc_6, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(6) },
-                      { C::instr_fetching_sel_op_dc_7, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(7) },
-                      { C::instr_fetching_sel_op_dc_8, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(8) },
-                      { C::instr_fetching_sel_op_dc_9, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(9) },
-                      { C::instr_fetching_sel_op_dc_10, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(10) },
-                      { C::instr_fetching_sel_op_dc_11, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(11) },
-                      { C::instr_fetching_sel_op_dc_12, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(12) },
-                      { C::instr_fetching_sel_op_dc_13, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(13) },
-                      { C::instr_fetching_sel_op_dc_14, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(14) },
-                      { C::instr_fetching_sel_op_dc_15, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(15) },
-                      { C::instr_fetching_sel_op_dc_16, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(16) },
-                      { C::instr_fetching_sel_op_dc_17, WIRE_INSTRUCTION_SPEC.at(w_opcode).op_dc_selectors.at(17) },
-                  } });
-        row++;
+    // We start from row 1 because we need a row of zeroes for the shifts.
+    uint32_t row = 1;
+
+    // Group and process events per bytecode_id.
+    // The trace is expecting all rows pertaining to a given bytecode_id to be contigous.
+    unordered_flat_map<BytecodeId, std::vector<size_t>> bytecode_id_to_events;
```

Comment:
I think we should/can avoid this but if we cannot I will then leave some comments on how to make this more efficient.